cluster-autoscaler/simulator/clustersnapshot/store/delta.go

```diff
@@ -428,24 +448,66 @@ func (snapshot *DeltaSnapshotStore) AddSchedulerNodeInfo(nodeInfo *schedulerfram
 	return nil
 }
 
+// setClusterStatePodsSequential sets the pods in cluster state in a sequential way.
+func (snapshot *DeltaSnapshotStore) setClusterStatePodsSequential(nodes []*apiv1.Node, nodeNameToIdx map[string]int, scheduledPods []*apiv1.Pod) error {
+	for _, pod := range scheduledPods {
+		if _, ok := nodeNameToIdx[pod.Spec.NodeName]; ok {
+			if err := snapshot.data.addPod(pod, pod.Spec.NodeName); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// setClusterStatePodsParallelized sets the pods in cluster state in parallel based on snapshot.parallelization value.
+func (snapshot *DeltaSnapshotStore) setClusterStatePodsParallelized(nodes []*apiv1.Node, nodeNameToIdx map[string]int, scheduledPods []*apiv1.Pod) error {
+	podsForNode := make([][]*apiv1.Pod, len(nodes))
+	for _, pod := range scheduledPods {
+		nodeIdx, ok := nodeNameToIdx[pod.Spec.NodeName]
+		if !ok {
+			continue
+		}
+		podsForNode[nodeIdx] = append(podsForNode[nodeIdx], pod)
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	workqueue.ParallelizeUntil(ctx, snapshot.parallelization, len(nodes), func(nodeIdx int) {
+		err := snapshot.data.addPodsForNode(podsForNode[nodeIdx], nodes[nodeIdx].Name)
+		if err != nil {
+			cancel(err)
+		}
+	})
+
+	return context.Cause(ctx)
+}
+
 // SetClusterState sets the cluster state.
 func (snapshot *DeltaSnapshotStore) SetClusterState(nodes []*apiv1.Node, scheduledPods []*apiv1.Pod, draSnapshot drasnapshot.Snapshot) error {
 	snapshot.clear()
 
-	knownNodes := make(map[string]bool)
-	for _, node := range nodes {
+	nodeNameToIdx := make(map[string]int, len(nodes))
```

Comment:
If you modified `addNode` to return `(*NodeInfo, error)` instead of just `error`, you could map names to infos directly.